---
title: "Genetic Associations of Susceptibility to Typhoid Fever"
author: "Mari Johnson"
date: "2022-10-12"
output: html_document
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE, eval = TRUE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(root.dir = '~/GWAS_22/gwas_final/merge/typhoid/assoc')

library(dplyr)
library(data.table)
library(ggplot2)
library(ggpubr)
library(qqman)
library(tidylog)
library(stringr)
library(ggrepel) 
library(tidyr)
library(tidylog)
library(stringi)
library(janitor)
library(stringr)
library(RColorBrewer)
library(forcats)
library(viridis)

```

### Aims:

-   Using imputed and cleaned data, associate genome-wide SNP frequency in our cohort with susceptibility/protection from typhoid fever following oral challenge (GWAS)
-   Association analysis can be modified to also look at regions of interest (CGAS)
-   Check Q-Q plot and genomic inflation values
-   Visualise GWAS significance with Manhattan plots
-   Interaction testing of covariates
-   Filter and format top hit data for downstream annotation

### Linear Regression GWAS models

Plink2 fits generalised linear mixed effects model:

y = GùõΩ + CùõΩ + e

-   Where, y is our response variable - ReSVinet score
-   G is our genotype dosage matrix
-   C is our covariate matrix
-   e is the error standard error (or residual) of our model estimate
-   ùõΩ are our model coefficients of determination. So, for each SNP we are modelling whether zero, one or two copies of an allele in our population, against our linear y variable, RSV severity.

We then test to see if there is a relationship between allele dosage and outcome, how strong the effect is, if it is significant

The null hypothesis is that there is no linear relationship. Therefore if;\
ùõÉ = 0 the allele shows no relationship with our phenotype score\
ùõÉ \> 0 the allele is associated with a more positive score\
ùõÉ \< 0 the allele is associated with a more negative score

The significance of this relationship is tested by calculating the T statistic:\
\
**t = b / SE** - **b** is the estimated value of linear slope or the coefficient of the predictor variable.\
- **SE~*b*~** represents the standard error of the coefficient estimation which can be estimated using the following formula:\
- **SE = S / ‚àöN** Where S represents the standard deviation and N represents the total number of data points

If the p-value that corresponds to t is less than some threshold (e.g. Œ± = .05); then we reject the null hypothesis and conclude that there is a statistically significant relationship between the predictor variable and the response variable

### Case-control analysis

-   General linear association model (GLM): Diagnosis
-   Covariates to include: PCs, Sex, <https://www.cog-genomics.org/plink/2.0/assoc>
-   MAF 0.1, SNPs have to be in at least 10% of the data i.e 25 participants

#### Enteric Fever

```{bash GWAS Enteric fever}
plink2 --bfile ~/GWAS_22/gwas_final/merge/enteric/QC/enteric.IBD \
--maf 0.1 \
--pheno ~/GWAS_22/gwas_final/meta/pheno_enteric.txt \
--pheno-name Diagnosed \
--covar ~/GWAS_22/gwas_final/meta/enteric_pcacovar.txt \
--covar-name age, sex, chall_vax, PC1, PC2, PC3, PC4, PC5 \
--covar-variance-standardize \
--glm \
--adjust \
--out enteric
```

#### Typhoid fever

```{bash GWAS Typhoid fever}
plink2 --bfile ~/GWAS_22/gwas_final/merge/typhoid/QC/typhoid2.IBD \
--maf 0.1 \
--pheno ~/GWAS_22/gwas_final/meta/pheno_typhoid.txt \
--pheno-name Diagnosed \
--covar ~/GWAS_22/gwas_final/meta/typhoid_pcacovar.txt \
--covar-name sex, age, chall_vax, PC1, PC2, PC3, PC4, PC5 \
--covar-variance-standardize \
--glm \
--adjust \
--out typhoid2
```

New PCs lamda (check using --adjust flag) = 1, no genomic inflation :)

### Genome-wide association: Results

**GWAS**

```{r Plot Assoc Results, eval = TRUE, cache=TRUE}
# Typhoid Assoc -----------------------------------------------------------------
test <- c("igg_2")

setwd("~/GWAS_22/gwas_final/merge/typhoid/vast")

# Logistic
#gwas <- fread(paste0(test, ".Diagnosed.glm.logistic.hybrid"))
# colnames(gwas) <- c("CHR","BP", "SNP","REF","ALT","A1", "FIRTH?", "TEST","OBS_CT","OR","LOG(OR)_SE","Z_STAT", "P","ERRCODE")
# gwas <- select(gwas, !c(`FIRTH?`,ERRCODE, OBS_CT)) 


#Linear
gwas <- fread(paste0(test, ".glm.linear"))

# Rename cols
colnames(gwas) <-
  c("CHR","BP", "SNP","REF","ALT","A1",
    "TEST","OBS_CT","BETA", "SE",
    "T_STAT", "P","ERRCODE"
  )

gwas <- filter(gwas, TEST == "ADD") 
gwas <- select(gwas, !c(ERRCODE, OBS_CT))

# Inspect top snps: ------------------------------------------------------------
tophits <- filter(gwas, P < 1e-4)

write.table(tophits, file = paste0(test, "_tophits.txt"), quote = F, row.names= F, sep = "\t")

write.table(gwas, file = paste0(test, "_gwas.txt"), quote = F, row.names= F, sep = "\t")
# delete og assoc output as very large

# QQ plot
# plot_dir <- c("~/GWAS_22/gwas_final/merge/enteric/assoc/plots/")
# png(file = (paste(plot_dir,test"_qq.png", sep = "")), width = 7, height = 7, units = "in", res = 300)
# qq(gwas$P)
# dev.off()
```

#### Manhattan plot

-   Can add labels from annotation later

```{r manhattan-setup}
#gwas <- paste0(test, "_gwas.txt")

gwas_data <- clean_names(gwas)

# Reduce comp time by removing dense plots at the bottom which overlap
sig_data <- gwas_data %>% 
  subset(p < 0.01)
nsig_data <- gwas_data %>% 
  subset(p >= 0.01) %>%
  group_by(chr) %>% 
  sample_frac(0.1) # Remove 1/10th of the data
gwas_data <- bind_rows(sig_data, nsig_data)

# Calculate base pair coordinates -----------------------------
data_cum <- gwas_data %>% 
group_by(chr) %>% 
  summarise(max_bp = max(bp)) %>% 
  mutate(bp_add = lag(cumsum(as.numeric(max_bp)), default = 0)) %>% 
  select(chr, bp_add)

gwas_data <- gwas_data %>% 
  inner_join(data_cum, by = "chr") %>% mutate(bp_cum = bp + bp_add)

gwas_data <- gwas_data %>%
select(-test,-bp_add)

axis_set <- gwas_data %>% 
  group_by(chr) %>% 
  summarize(center = mean(bp_cum))

ylim <- gwas_data %>% 
  filter(p == min(p)) %>% 
  mutate(ylim = abs(floor(log10(p))) + 2) %>% 
  pull(ylim)
sig <- 1e-5 # Top = 9.9e-5
```

```{r basic-manhattan}
# Basic Plot ----------------------------------------------------
setwd("~/GWAS_22/gwas_final/merge/typhoid/vast")
plot_dir <- plot_dir <- paste0(getwd(),"/plots/")
png(file = (paste0(plot_dir,test,"-man.png")), width = 9, height = 6, units = "in", res = 300)
ggplot(gwas_data,
       aes(
         x = bp_cum,
         y = -log10(p),
         color = as_factor(chr),
         size = -log10(p)
       )) +
  geom_hline(yintercept = -log10(sig),
             color = "grey40",
             linetype = "dashed") +
  geom_point(alpha = 0.8, size = 1.3) +
  scale_x_continuous(label = axis_set$chr, breaks = axis_set$center) +
  scale_y_continuous(expand = c(0, 0), limits = c(0, 8)) +
  scale_color_manual(values = rep(c("#3d8e7f", "#93d5b2"), 22)) +
  scale_size_continuous(range = c(0.5, 3)) +
  labs(x = "Chromsome",
       y = "log(p)\n-") + theme_minimal() +
  theme(
    legend.position = "none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  )
dev.off()
```

### FcR association testing

Ensembl coords +/- 10 KB Starting from FcGR2a Chromosome 1: 161,505,430 To FcGR2b Chromosome 1: 161,678,654

```{bash GWAS FCGR}
plink2 --bfile ~/GWAS_22/gwas_final/merge/typhoid/QC/typhoid2.IBD \
--maf 0.1 \
--chr 1 \
--from-bp 161505430 \
--to-bp 161688654 \
--pheno ~/GWAS_22/gwas_final/meta/pheno_typhoid.txt \
--pheno-name Diagnosed \
--covar ~/GWAS_22/gwas_final/meta/typhoid_pcacovar.txt \
--covar-name sex, chall_vax, PC1, PC2, PC3, PC4, PC5 \
--covar-variance-standardize \
--glm \
--adjust \
--out fcgr_assoc

 #TODO need to check LD patterns of surrounding genes in the locus along with gwas assoc data
 # I.e is there a more sig snp in a nearby LD gene that is responsible for the association
```

### FcaR association testing

Ensembl coords GRCh38.p13: 19: 54,874,235-54,891,420

```{bash Fcar}
plink2 --bfile ~/GWAS_22/gwas_final/merge/typhoid/QC/typhoid2.IBD \
--maf 0.1 \
--chr 19 \
--from-bp 54864235 \
--to-bp 54901420 \
--pheno ~/GWAS_22/gwas_final/meta/pheno_typhoid.txt \
--pheno-name Diagnosed \
--covar ~/GWAS_22/gwas_final/meta/typhoid_pcacovar.txt \
--covar-name sex, chall_vax, PC1, PC2, PC3, PC4, PC5 \
--covar-variance-standardize \
--glm \
--adjust \
--out fcar_assoc

 #TODO need to check LD patterns of surrounding genes in the locus along with gwas assoc data
 # I.e is there a more sig snp in a nearby LD gene that is responsible for the association
```

### VAST antibody titre and function GWAS

```{bash VAST-GWAS}

# Update pheno and file names as required
cd ~/GWAS_22/gwas_final/merge/typhoid/vast
plink2 --bfile vast \
--maf 0.1 \
--pheno pheno_vast.txt \
--pheno-name vi_ig_g_titre \
--covar covar_vast.txt \
--covar-name sex, age, chall_vax, PC1, PC2, PC3, PC4, PC5 \
--covar-variance-standardize \
--glm \
--adjust \
--out igg_2
```

```{bash VAST-GWAS}

# Update pheno and file names as required
cd ~/GWAS_22/gwas_final/merge/typhoid/vast
plink2 --bfile vast \
--maf 0.1 \
--pheno pheno_vast.txt \
--pheno-name adnp \
--covar covar_vast.txt \
--covar-name sex, age, chall_vax, PC1, PC2, PC3, PC4, PC5 \
--covar-variance-standardize \
--glm \
--adjust \
```




##### VAST FcgR/Fcar
Q: are snps in the FcGR region affecting vaccine responses FcR

```{bash VAST-FcGR}
# Update pheno and file names as required
plink2 --bfile ~/GWAS_22/gwas_final/merge/typhoid/QC/typhoid2.IBD \
--maf 0.15 \
--chr 1 \
--from-bp 161505430 \
--to-bp 161688654 \
--pheno ~/GWAS_22/gwas_final/meta/pheno_vast.txt \
--pheno-name Diagnosed \
--covar ~/GWAS_22/gwas_final/meta/typhoid_pcacovar.txt \
--covar-name sex, chall_vax, PC1, PC2, PC3, PC4, PC5 \
--covar-variance-standardize \
--glm \
--adjust \
--out vast/vast.fcgr
```
```{bash VAST-FcaR}
# Update pheno and file names as required
plink2 --bfile ~/GWAS_22/gwas_final/merge/typhoid/QC/typhoid2.IBD \
--maf 0.15 \
--chr 19 \
--from-bp 54864235 \
--to-bp 54901420 \
--pheno ~/GWAS_22/gwas_final/meta/pheno_vast.txt \
--pheno-name Diagnosed \
--covar ~/GWAS_22/gwas_final/meta/typhoid_pcacovar.txt \
--covar-name sex, chall_vax, PC1, PC2, PC3, PC4, PC5 \
--covar-variance-standardize \
--glm \
--adjust \
--out vast/vast.fcar

# vast_pheno.txt for all ab vars, pheno_vast.tx for diagnosis TODO:fix these later

```

#### Previous GWAS SNPs

TLR4:
9: 117,693,175-117,734,735 
Not significant P = 0.1

VAC14
16: 70,677,439-70,811,160
most sig is 0.000210881, (2e-4), not genome wide

PARK2,
CHR 6 161337417-162737775
P = 0.0152076

CFTR,
Chromosome 7: 117277120-117725971 
P = 0.0236181

CAPN,
Chromosome 2: 31,163,056-31,243,858
NS, P = 0.08


```{bash TLR}
plink2 --bfile ~/GWAS_22/gwas_final/merge/typhoid/QC/typhoid2.IBD \
--maf 0.1 \
--chr 2 \
--from-bp 161337417 \
--to-bp 162737775  \
--pheno ~/GWAS_22/gwas_final/meta/pheno_typhoid.txt \
--pheno-name Diagnosed \
--covar ~/GWAS_22/gwas_final/meta/typhoid_pcacovar.txt \
--covar-name sex, chall_vax, PC1, PC2, PC3, PC4, PC5 \
--covar-variance-standardize \
--glm \
--adjust \
--out park2
```

### Candiate gene association: Results

**CGAS**









```{r CGAS Results, eval = TRUE, cache=TRUE}
# Typhoid Assoc -----------------------------------------------------------------

gwas <- fread("vast.fcgr.enteric.glm.linear")

gwas <-
  filter(gwas, TEST == "ADD") 

# Rename cols - Linear
colnames(gwas) <-
  c("CHR","BP", "SNP","REF","ALT","A1",
    "TEST","OBS_CT","BETA", "SE",
    "T_STAT", "P","ERRCODE"
  )

gwas <- select(gwas, !c(ERRCODE, OBS_CT)) 

# Inspect top snps: ------------------------------------------------------------
tophits <- filter(gwas, P < 0.05)

write.table(tophits, file = "tophits_fcgr_enteric.txt", quote = F, row.names= F, sep = "\t")

write.table(gwas, file = "vast_fcgr_enteric.txt", quote = F, row.names= F, sep = "\t")
# delete og assoc output as very large
```

Summary: VAST enteric/ADNOB

- ADNOB, sig FCGR snps, 3 FcAR (just)
- enteric, no sig FcAR snps, lots of sig fcar snps
- Diagnosis

Summary: FcRs challenge

- FcGR snps assoc with 


### Interaction testing

Warning message: \--glm remaining control count is less than 10x predictor count for phenotype 'Diagnosed'\
As I now have a much smaller sample size and more covariates I have 10 covariates which will be less than 1/10th of our n 153.

I am already thinking there is probably issues with linear dependencies based on the covar, i.e with dose (poss concat this with challenge).\
To many covars and too small an n can overfit model.\
Therefore test for significance of these factors in linear model.\
Remove least significant covars.

-   covar standardise necessary, but does pass VIF test (=2)

```{bash GWAS Covar checking}
plink2 --bfile ~/GWAS_22/gwas_final/merge/typhoid/QC/typhoid2.IBD \
--pheno ~/GWAS_22/gwas_final/meta/pheno_typhoid.txt \
--pheno-name Diagnosed \
--covar ~/GWAS_22/gwas_final/meta/typhoid_pcacovar.txt \
--covar-name sex, chall_vax, PC1, PC2, PC3, PC4, PC5 \
--glm genotypic interaction \
--covar-variance-standardize \
--out covar_checks/check
```

```{r Interaction test}
interact <- fread("covar_checks/check.Diagnosed.glm.logistic.hybrid")
colnames(interact) <-
  c("CHR","BP", "SNP","REF","ALT","A1",
    "FIRTH?", "TEST","OBS_CT","OR","LOG(OR)_SE",
    "Z_STAT", "P","ERRCODE")

# Remove unnecessary columns
interact <- interact %>% 
  select(!c(ALT, OBS_CT,`FIRTH?`,`LOG(OR)_SE`, Z_STAT, OR)) 

interact$TEST <- as.factor(interact$TEST)
levels(interact$TEST)
VIF_high <- filter(interact, ERRCODE == "VIF_TOO_HIGH")
table(VIF_high$TEST) # 998,140 SNPs have error codes 

VIF_ADD <- filter(interact, str_detect(TEST, "ADD")==TRUE)
levels(VIF_ADD$TEST)

# 1947595 - VIF too high, are these for low sig snps?
# Filter for top SNPS? Then look at interaction factor sig
# Combine vaccine and challenge into one factor

# Just vaccine 
vax <- interact[(interact$TEST == "chall_vax"),]
vax %>%
  group_by(SNP) %>%
  ggplot(aes(
    x = TEST,
    y = -log10(P),
    fill = TEST
  )) +
  geom_boxplot() +
  theme_bw() +
  labs(x = "\nCovariate", title = "Interaction Tests") +
  theme(legend.position = "none")

# ADD
geno <- interact %>% filter(TEST == "ADD")
geno <- filter(geno, P <1e-2) #topsnps = <10e-4
snps <- geno$SNP
top <- filter(ADD, SNP %in% snps)

# glm is fitting a model per snp, therefore get an overview of covars
tiff(file = "interaction-snps.tiff",
     res = 720,
     width = 7500, height = 5000,
     compression = "lzw")
top %>%
  group_by(SNP) %>%
  ggplot(aes(
    x = TEST,
    y = -log10(P),
    fill = TEST
  )) +
  geom_boxplot() +
  theme_bw() +
  labs(x = "\nCovariate", title = "Interaction Tests") +
  theme(legend.position = "none")
dev.off()

```
