---
title: "SNP Annotation"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, warning = FALSE, message = TRUE)
knitr::opts_knit$set(root.dir = '~/GWAS_22/gwas_final/merge/typhoid/assoc/nexus/adnob')
```

### Aims

-   Upload SNP list to SNPNexus/VEP
-   Results in various annotation files
-   Download snp annotation results as .txt files and read into R
-   For candidate gene analysis prioritise and filter SNPs with GTex tissue expression, coding snps etc.

##### Load Packages:

```{r Packages}
library(dplyr)
library(data.table)
library(ggplot2)
library(ggrepel)
library(tidylog)
library(stringr)
library(stringi)
library(openxlsx)
library(RColorBrewer)
library(forcats)
library(qqman)
library(janitor)
```

#### Write nexus snp list

Upload data to SNP nexus to get genome annotation <https://www.snp-nexus.org/v4/>

```{r format Nexus snp list}
tophits <- fread("tophits/tophits_adnob_age.txt")
nexus <- select(tophits, CHR, BP, REF, ALT)
nexus <- mutate(nexus, strand = rep(1, length(nexus$CHR)))
nexus <- mutate(nexus, Type = rep("Chromosome", length(nexus$CHR)))
nexus <- select(nexus, Type, CHR, BP, REF, ALT, strand)

names(nexus) <-
  c("Type", "Id", "Position", "Alelle1", "Allele2", "Strand")

write.table(nexus, "adnob2_nexus.txt", row.names = F, col.names = F, quote = F, sep = " ")
#Then need to find what these snps are in LD with
```

Read nexus .txt files from assoc directory into R

#### Format Annotation Data

```{r get-nexusdata, message=FALSE}
# Read in data
temp = list.files(pattern = ".txt")
myfiles = lapply(temp, fread) #read all files in temp
trim <- str_extract(temp, ".{4,}?")
names(myfiles) <- trim

#Unlist files and save to global environment
list2env((myfiles), envir = .GlobalEnv)
tophits <- toph
```
```{r Format Annotation data ID, message=TRUE}
#ID file -----------------------------------------------------------------------

ID <- near #contains file of nearest gene to the snps
rm(near)
ID <- ID[,1:8]

#Predicted ensembl function ----------------------------------------------------
coding_muts <-
  filter(ense,`Predicted Function` == "coding")

ense <-
  ense %>%
  dplyr::select(`Variation ID`, Variant, Symbol, `Predicted Function`) %>%
  distinct()
#Lots of duplicate entries due to contigs I think

introns <-
  ense %>% filter(str_detect(ense$`Predicted Function`, "intron"))

no_introns <-
  ense %>% filter(!str_detect(ense$`Predicted Function`, "intron"))

#collapse by ense by ID
ense <- ense %>%
  group_by(`Variation ID`, Symbol) %>%
  summarise(`Predicted Function` = toString(`Predicted Function`))

main <- full_join(ID, ense, by = "Variation ID")


#Regulatory Elements -----------------------------------------------------------
#road <- fread("roadmap_mari.johnson@lmh.ox.ac.uk.txt")
road$Epigenome <- as.factor(road$Epigenome)
#levels(road$Epigenome)
# Contains many tissue eQTLs
# Filter for immune and lung tissue expression
road <-
  road %>%
  filter(str_detect(road$Epigenome, "B cell|CD4|CD8|CD34|T cell|T helper|Monocyte|neutrophil|myeloid|Natural Killer|Spleen|macrophage|colon|Intestine")==TRUE)

# Collapse by Histone feautre
road <- road %>%
  group_by(`Variation ID`,
           `Chromosome`,
           `Feature Type Class`) %>%
  summarise(Tissue_Express = toString(Epigenome))

road$Tissue_Express <- str_remove_all(road$Tissue_Express, ("Roadmap"))
road$Tissue_Express <- str_remove_all(road$Tissue_Express, ("PB"))
road <- filter(road, Tissue_Express != "Fetal Intestine Small")
main <- left_join(main, road, by = "Variation ID")


# GWAS public data -------------------------------------------------------------

gwas <- gwas %>%
  select(`Variation ID`, Trait, `p-Value`,`Risk Allele`, PubMed)
```

```{r Join tables, message=TRUE}
# Join tables ------------------------------------------------------------------

tophits <- select(tophits, -c(TEST))

# Format SNP - VAR ID for joining
# chr1:25346917:G:A > chr1:111418204:G/A:1

tophits <- rename(tophits, `Variation ID` = SNP)
tophits$`Variation ID` <- gsub("((?:[^//:]+//:){3}[^//:]+)//:", "\\1i", tophits$`Variation ID`)

main$`Variation ID`<- str_remove_all(main$`Variation ID`, ("\\:1$"))


# Replace the third (or last) instance of :
# with // / slash
replN <- function(x,fn,rp,n) {
    sel <- rep(fn, n*length(rp))
    sel[seq_along(rp)*n] <- rp
    regmatches(x, gregexpr(fn, x)) <- list(sel)
    x
}

tophits$`Variation ID` <- replN(tophits$`Variation ID`, fn=":", rp=c("/"), n=3)
#tophits$`Variation ID` <- paste(tophits$`Variation ID`, ":1", sep = "")
tophits <- dplyr::rename(tophits, position = BP)

gwas <- clean_names(gwas)
path <- clean_names(path)
road <- clean_names(road)
ID <- clean_names(ID)
main <- clean_names(main)
main <- main %>% select(!ends_with("_y"))
names(main) <- str_replace_all(names(main),"_x","")

tophits <- clean_names(tophits)
main <- left_join(tophits, main, by = c("variation_id", "position"))


spread <- list('tophits' = main,
               'public_gwas' = gwas,
               'pathway' = path)

write.xlsx(spread, file = "typhoid_gwas.xlsx")
write.csv(main, file = "GWAS_summary.data.csv", row.names = F)

```
#### Labelled Manhattan Plots

```{r Labelled-manhattan}
gwas_data <- clean_names(fread("vast_gwas.txt"))
#main <- read.csv(file = "GWAS_summary.data.csv")

# Reduce comp time by removing dense plots at the bottom which overlap
sig_data <- gwas_data %>% 
  subset(p <= 0.01)
nsig_data <- gwas_data %>% 
  subset(p >= 0.01) %>%
  group_by(chr) %>% 
  sample_frac(0.1) # Keep 8% of nsig data
gwas_data <- bind_rows(sig_data, nsig_data)

# Calculate base pair coordinates -----------------------------
data_cum <- gwas_data %>% 
group_by(chr) %>% 
  summarise(max_bp = max(bp)) %>% 
  mutate(bp_add = lag(cumsum(as.numeric(max_bp)), default = 0)) %>% 
  select(chr, bp_add)

gwas_data <- gwas_data %>% 
  inner_join(data_cum, by = "chr") %>% 
  mutate(bp_cum = bp + bp_add)

gwas_data <- gwas_data %>%
select(-test,-bp_add)

axis_set <- gwas_data %>% 
  group_by(chr) %>% 
  summarize(center = mean(bp_cum))

ylim <- gwas_data %>% 
  filter(p == min(p)) %>% 
  mutate(ylim = abs(floor(log10(p))) + 2) %>% 
  pull(ylim)
sig <- 1e-5 # Top = 9.9e-5
```
```{r highlight genes}
# Highlight genes --------------------------------------------------------------

# Leave public hits for now, didn't seem v relevant
#public <- read.xlsx(xlsxFile = "typhoid_gwas.xlsx", sheet = "public_gwas")
#eqtl <- load(file = "~/RSV/data/transcriptomics/eQTL/eqtl_results.RData")

# Just remove dups for now
main <- main[(stri_duplicated(main$variation_id)==FALSE),]

# Update var_id format 
main$snp <- str_replace_all(main$variation_id, "\\/", "\\:") 
#public$snp <- str_replace_all(public$variation_id, "\\/", "\\:") 
#public$snp <- str_remove_all(public$snp, "(\\:1)$")

# variation_ids to highlight ---------------------------------------------------
# filter so grouped by gene and then take the min val

DT <- data.table(main)
top <- DT[, .SD[which.min(p)], by = overlapped_gene]
top <- top %>% filter(overlapped_gene != "None")
top <- top %>% filter(overlapped_gene != "none")
top_snps <- top$snp

#public_snps <- public$snp
#cis_snps <- cis$snps
#trans_snps <- trans$snps

# Add highlight and annotation information
gwas_data <- gwas_data %>%
  mutate(highlight_top = ifelse(snp %in% top_snps, "yes", "no"))
  # mutate(highlight_pub = ifelse(snp %in% public_snps, "yes", "no"))
  # mutate(highlight_cis = ifelse(snp %in% cis_snps, "yes", "no")) %>%
  # mutate(highlight_trans = ifelse(snp %in% trans_snps, "yes", "no"))

# Get gene names
gene <- select(main, snp, overlapped_gene)
gwas_data <- left_join(gwas_data, gene, by = "snp")

#gwas_data <- left_join(gwas_data, public, by = "snp") %>% select(-variation_id)
#gwas_data <- gwas_data %>% select(!ends_with(".y"))
names(gwas_data) <- str_replace_all(names(gwas_data),".x","")
#TODO turn name cleaning after join into a function

#eqtls <- rbind(cis, trans) %>% rename(snp = snps)
#eqtls <- filter(eqtls, !stri_duplicated(gene))
#eqtl_gene <- select(eqtls, snp, gene)
#gwas_data <- left_join(gwas_data, eqtl_gene, by = "snp")

```
```{r plot-manhattan}
# Plot with labelled eqtl genes ------------------------------------------------
plot_dir <- c("~/GWAS_22/gwas_final/merge/typhoid/assoc/plots/")

png(file = (paste(plot_dir,"top-genes-adnob-man.png", sep = "")), width = 10, height = 6, units = "in", res = 300)
ggplot(gwas_data,
       aes(
         x = bp_cum,
         y = -log10(p),
         color = as_factor(chr),
         size = -log10(p)
       )) +
  geom_hline(yintercept = -log10(1e-5),
             color = "grey34",
             linetype = "dashed") +
  geom_hline(yintercept = -log10(1e-8),
             color = "grey80",
             linetype = "dashed") +
  geom_point(alpha = 0.8, size = 1.3) +
  scale_x_continuous(label = axis_set$chr, breaks = axis_set$center) +
  scale_y_continuous(expand = c(0, 0), limits = c(0, 8.5)) +
  scale_color_manual(values = rep(c("#3d8e7f", "#93d5b2"), 22)) +
  scale_size_continuous(range = c(0.5, 3)) +
  geom_label_repel(
    data = subset(gwas_data, highlight_top == "yes"),
    aes(label = overlapped_gene),
    size = 2.5, color = "#202936", nudge_y = 0.5, nudge_x = 0.1
  ) +
    theme_minimal() +
  theme(
    legend.position = "none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  ) +
  labs(x = "Chromsome", title = "Genetic associations of Vi-ADNOB responses following typhoid vaccination")
dev.off()

save.image(file = "adnob.RData")

# Colour palette
# #fcf2c2 light colour, #e06252 light red, #202936 dark blue, #D28423 og pale yellow/orange
# +
#   geom_label_repel(
#     data = subset(gwas_data, highlight_pub == "yes"),
#     aes(label = trait),
#     size = 2.5, color = "#AD3A8A", nudge_y = -0.3, nudge_x = 2) +
#   labs(x = "Chromsome",
#        y = "-log10(p)") 

```

#### GWAS combined tophits

- check for common genes/variants

