---
title: "SNP Annotation"
author: "Mari Johnson"
date: "23/03/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(root.dir = '~/GWAS_22/new_gwas/Post-imp/Merged/Assoc/')
```
##### Load Packages:
```{r Packages, message=FALSE}
library(data.table)
library(tidylog)
library(dplyr)
library(stringr)
```
### Annotate FcY SNPs
24/03/22
- Redo annotation with **just the typhoid** snps
- Update to include general tophits as well as fcr snps at the same time

##### **Plan**
- Make annotation file for https://www.snp-nexus.org/v4/guide/
- Remake enteric bim file without the rs IDs (*fix later)
- Use snp nexus online tool for now which will also give rs IDs as an output 
- Later integrate data from USCS using mySQL
 
### Write nexus snp list
```{r write nexus snp list}

fcy_sig <- fread("T_fcy.txt")
fcar_sig <- fread("T_fcar.txt")
T_tophits <- fread("T_tophits.txt")

sigsnps <- rbind(fcar_sig, fcy_sig, T_tophits)#From assoc analysis

tophits <- fread("tophits.txt")
nexus <- select(sigsnps, CHR, BP, REF, ALT) #957
nexus <- mutate(nexus, strand = rep(1, 111))
nexus <- mutate(nexus, Type = rep("Chromosome", 111))
nexus <- select(nexus, Type, CHR, BP, REF, ALT, strand)
names(nexus) <- c("Type","Id","Position","Alelle1","Allele2","Strand")
write.table(nexus, "annotation/nexus/nexus.txt", row.names = F, col.names = F, quote = F, sep = " ")
#Then need to find what these snps are in LD with 
```

### Format Annotation data

- Upload SNP list to SNPNexus
- Results in various annotation files
- Download snp annotation results as .txt files and read into R 
```{r Load Data, message=FALSE}
#Read in data
temp = list.files(pattern="*.txt")
myfiles = lapply(temp, fread)
#add names to files  - names(myfiles) <- temp
trim <- str_extract(temp, ".{4,}?")
names(myfiles) <- trim
#Unlist files
list2env((myfiles), envir = .GlobalEnv)
```

#### Format Data
```{r Format Data, message = TRUE}
#ID file ####
ID <- gen_ #contains file of nearest gene to the snps
rm(gen_)
ID <- ID[,1:8]

#Predicted ensembl function ####
ense <- ense %>% dplyr::select(`Variation ID`, Variant, Symbol, `Predicted Function`) %>% distinct()
#Lots of duplicate entries due to contigs I think
no_introns <- ense %>% filter(`Predicted Function` != "intronic")

introns <- ense %>% filter(str_detect(ense$`Predicted Function`, "intron"))
no_introns <- ense %>% filter(!str_detect(ense$`Predicted Function`, "intron"))

#collapse by ense by ID
functions <- ense %>%
  group_by(`Variation ID`, Symbol) %>%
  summarise(`Predicted Function` = toString(`Predicted Function`))
  
main <- full_join(ID, functions, by = "Variation ID")
#Matched 485 rows

#Regulatory Elements ####
#reg elements downloaded from roadmap project via snpnexus 
road <- fread("~/GWAS_22/new_gwas/Post-imp/Merged/Assoc/annotation/nexus/roadmap_Typhoid.txt")
str(road)
road$Epigenome <- as.factor(road$Epigenome)
levels(road$Epigenome)
#Contains many tissue eQTLs, #Filter for immune and lung tissue expression
epi <- road %>% filter(str_detect(road$Epigenome, "B cell|CD4|CD8|T cell|Monocyte|Neutrophil|myeloid|Natural Killer|Spleen|macrophage|Small Intestine|sigmoid colon"))
#Collapse by SNP
epi <- epi %>%
  group_by(`Variation ID`, Chromosome, `Feature Type Class`, `Feature Type`) %>%
  summarise(Epigenome = toString(Epigenome))
#cat Feature Type and Epigenome to one column
epi <- epi %>% group_by(`Variation ID`, Chromosome, `Feature Type Class`) %>% mutate(Epigenome_Feature = paste(`Feature Type`, `Epigenome`, sep = ": ")) %>% select(-`Feature Type`, -Epigenome)
main <- full_join(main, epi, by = "Variation ID")

#CADD ####
cadd <- cadd %>% select(`Variation ID`,`Raw Score`,PHRED)
main <- full_join(main, cadd, by = "Variation ID")

#GWAS public data ####
gwas <- gwas%>% select(-`Risk Allele`, -Region, -Platform)
main <- full_join(main, gwas, by = "Variation ID")
main <- distinct(main)

#Glm association results ####
assoc <- sigsnps %>% dplyr::select(CHR,BP,SNP,OR,`LOG(OR)_SE`,P)
main <- main %>% dplyr::rename(CHR=Chromosome.x, BP=Position)
main <- left_join(assoc, main, by = c("CHR","BP"))
main <- main %>% select(-Chromosome.y,-Chromosome,-`Variation ID`)

#Save table
write.csv(main, file = "~/GWAS_22/new_gwas/Post-imp/Merged/Assoc/annotation/nexus/GWAS_summary.data.csv", row.names = F)

#Just FCG receptors
Fcs <- main %>% filter(str_detect(main$Symbol, "FC"))

#collapse by rsID
#Cat Feature Class
epi <- epi %>%
  group_by(`Variation ID`, Chromosome, `Feature Type Class`, `Feature Type`) %>%
  summarise(Epigenome = toString(Epigenome))
#cat Feature Type and Epigenome to one column
Fcs <- Fcs %>% group_by(SNP) %>% mutate(Epigenome_Feature_Class = paste(`Feature Type Class`, `Epigenome_Feature`, sep = ": ")) %>% select(-`Feature Type Class`, -Epigenome_Feature)

colnames <- str_remove(names(Fcs),"Epigenome_Feature_Class")
rm(test)
Fcs <-Fcs %>% group_by(across(c(-Epigenome_Feature_Class))) %>%
  summarise(Epigenome_Feature_Class = toString(Epigenome_Feature_Class))
write.csv(Fcs, file = "~/GWAS_22/new_gwas/Post-imp/Merged/Assoc/annotation/nexus/FcyRs_assoc_annotate.csv", row.names = F)
```

```{r gtEX, echo = FALSE}
#query for whole blood
```
##### FcGR Notes:
- Top SNP associated with enteric fever, chr1:161630802:A/G:1, has come out as significant for GWAS, reduces IgG levels in Iceland study
- In gtEX database it increases FcGR2B expression - which is an inhibitory receptor
- FcGR2B, is a Low affinity receptor. Involved in a variety of effector and regulatory functions such as phagocytosis of immune complexes and modulation of antibody production by B-cells. Binding to this receptor results in down-modulation of previous state of cell activation triggered via antigen receptors on B-cells (BCR), T-cells (TCR) or via another Fc receptor. Isoform IIB1 fails to mediate endocytosis or phagocytosis. Isoform IIB2 does not trigger phagocytosis.

#### FcaR notes

- Top Fcar snp associated with protection
- This snp  has been shown to increased kir expression whole blood, can recognise bacterial epitopes

### Next steps

- Plot annotation of snps using tracks
- Update fcgr, fcar, and top snps names in plink using an annotation file from snp nexus (eventually get sql query to work)
- Plot typhoid fever by genotype (find old scripts for this)
- Run eqtl analysis of top snps with gene expression following challenge