---
title: "Pre-Imputation Steps"
author: "Mari Johnson"
date: "07/03/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(eval = FALSE)
knitr::opts_chunk$set(engine.opts='-l')
```

## Introduction 

This script includes preparation steps for QC and formatting GWAS microarray data for imputation\
There are 3 eneteric GWAS experiments (across 6 studies) that need to be QC'd and merged;\
- T1 T2
- P1 and Tyger
- VAST and PATCH

Raw data was saved in PLINK1 format, updated to binary format (fam, bed, bim)

VAST_PATCH files have updated IDs which were saved as new bed files\
See ID_updating/New_IDs.R for details\
File Tracker.txt also has details of file types\

## Aims 

- To ensure all studies have the same RS IDs and genome version
- QC data for duplicates, A > T SNPs etc.
- Merge QC'd Study Data
- Format data into vcf.gz files for uploading to Topmed Imputation Server

## Set Up 

```{r Library}
setwd("~/GWAS_22/new_gwas")
library(data.table)
library(stringr)
library(dplyr)
library(tidylog)
library(stringi)
```
```{bash Set up folders}
cd ~/GWAS_22/new_gwas
#still need to troubleshoot setting bashrc profile to run in code chunks engine opts - l didn't work,
#Have moved plink2 to ~/bin which is in the markdown default shell PATH
```

```{bash Make folders}
mkdir ~/GWAS_22/new_gwas/updated_snps ;
mkdir ~/GWAS_22/new_gwas/updated_snps/VAST ;
mkdir ~/GWAS_22/new_gwas/updated_snps/T1 ;
mkdir ~/GWAS_22/new_gwas/updated_snps/Tyger
```

## VAST SNP ID formatting

```{r Liftover format BED file}
setwd("~/GWAS_22/new_gwas/VAST_PATCH")
#Download GSA BED file https://emea.support.illumina.com/downloads/infinium-global-screening-array-v3-0-support-files.htm
#check file and remove header
BED <- fread("GSA-24v3-0_A1.bed")  
write.table(file="GSA_BED.txt", BED, sep = "\t", quote=F, row.names = F, col.names = F)

#Liftover go online to https://genome.ucsc.edu/cgi-bin/hgLiftOver
#Submit file at the bottom of the page
#output:
#Successfully converted 625337 records
#Conversion failed on 2003 records.
#Download as VAST_38.bed
```

```{bash Updating Liftover SNPs}

cd ~/GWAS_22/new_gwas/VAST_PATCH
#Exctract mapped variants snp IDs
awk '{print $4}' VAST_38.bed > VAST_38_rm.snps
#Extract updated positions
awk '{print $4, $3}' VAST_38.bed > VAST_38_rm.pos
#Update plink file with SNPs that we could extract, and update the snp coordinates using update map function
plink2 --bfile VAST_PATCH_2 --make-pgen --out VP
plink2 --pfile VP --extract VAST_38_rm.snps --update-map VAST_38_rm.pos --sort-vars --make-pgen --out VAST_remap
plink2 --pfile VAST_remap --make-bed --out VAST_remap #bfile version
```
## T1 T2 formatting

```{r Clean T1 Bed files}
setwd("~/GWAS_22/new_gwas/T1T2")

#Load SNP info
bim<-fread("T1T2.bim")

#Filter SNPs in autosomal chromosomes")
non_auto <- bim %>% filter(V1 == "0" | V1 == "24")
#keep only SNPS containing rs IDs #
rs_bim <- bim %>% filter(!grepl("rs", V2))
#Remove duplicates 
duplicated_snps <- bim[duplicated(paste(bim$V1,bim$V4)),]

#remove snps where A > T, C > G
#PLINK can't tell what strand to merge/read these SNPs from
atcg_snps <-  bim[((bim$V5 == "A") & (bim$V6 == "T")) |
                         ((bim$V5 == "T") & (bim$V6 == "A")) |
                         ((bim$V5 == "C") & (bim$V6 == "G")) |
                         ((bim$V5 == "G") & (bim$V6 == "C"))]
#Exclusion list
exclude <- rbind(non_auto, duplicated_snps, atcg_snps, rs_bim)
write.table(file="exclude_snps_rm.txt", exclude, sep = "\t", quote=F, row.names = F, col.names = F)
system("plink --bfile T1T2 --exclude exclude_snps_rm.txt --make-bed --out T1T2_rm")
#error invalid centimorgan position on line 58721 of T1T2.bim.
#use Plink1.9 to convert for now, error reported
```


```{bash Make T1 Bed files}
cd ~/GWAS_22/new_gwas/T1T2
awk '{print "chr" $1, $4 -1, $4, $2 }' T1T2_rm.bim > T1T2_rm_lift

#https://genome.ucsc.edu/cgi-bin/hgLiftOver

```

