---
title: "Pre-Imputation Steps"
author: "Mari Johnson"
date: "07/03/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(eval = FALSE)
knitr::opts_chunk$set(engine.opts='-l')
```

## Introduction 

This script includes preparation steps for QC and formatting GWAS microarray data for imputation\
There are 3 eneteric GWAS experiments (across 6 studies) that need to be QC'd and merged;\
- T1 T2
- P1 and Tyger
- VAST and PATCH

Raw data was saved in PLINK1 format, updated to binary format (fam, bed, bim)

VAST_PATCH files have updated IDs which were saved as new bed files\
See ID_updating/New_IDs.R for details\
File Tracker.txt also has details of file types\

## Aims 

- To ensure all studies have the same RS IDs and genome version
- QC data for duplicates, A > T SNPs etc.
- Merge QC'd Study Data
- Format data into vcf.gz files for uploading to Topmed Imputation Server

## Set Up 

```{r Library}
setwd("~/GWAS_22/new_gwas")
library(data.table)
library(stringr)
library(dplyr)
library(tidylog)
library(stringi)
```
```{bash Set up folders}
cd ~/GWAS_22/new_gwas
#still need to troubleshoot setting bashrc profile to run in code chunks engine opts - l didn't work,
#Have moved plink2 to ~/bin which is in the markdown default shell PATH
```

```{bash Make folders}
mkdir ~/GWAS_22/new_gwas/updated_snps ;
mkdir ~/GWAS_22/new_gwas/updated_snps/VAST ;
mkdir ~/GWAS_22/new_gwas/updated_snps/T1 ;
mkdir ~/GWAS_22/new_gwas/updated_snps/Tyger
```

## VAST SNP ID formatting

```{r Liftover format BED file}
setwd("~/GWAS_22/new_gwas/VAST_PATCH")
#Download GSA BED file https://emea.support.illumina.com/downloads/infinium-global-screening-array-v3-0-support-files.htm
#check file and remove header
BED <- fread("GSA-24v3-0_A1.bed")  
write.table(file="GSA_BED.txt", BED, sep = "\t", quote=F, row.names = F, col.names = F)

#Liftover go online to https://genome.ucsc.edu/cgi-bin/hgLiftOver
#Submit file at the bottom of the page
#output:
#Successfully converted 625337 records
#Conversion failed on 2003 records.
#Download as VAST_38.bed
```

```{bash Updating Liftover SNPs}

cd ~/GWAS_22/new_gwas/VAST_PATCH
#Exctract mapped variants snp IDs
awk '{print $4}' VAST_38.bed > VAST_38_rm.snps
#Extract updated positions
awk '{print $4, $3}' VAST_38.bed > VAST_38_rm.pos
#Update plink file with SNPs that we could extract, and update the snp coordinates using update map function
plink2 --bfile VAST_PATCH_2 --make-pgen --out VP
plink2 --pfile VP --extract VAST_38_rm.snps --update-map VAST_38_rm.pos --sort-vars --make-pgen --out VAST_remap
plink2 --pfile VAST_remap --make-bed --out VAST_remap #bfile version
```
## T1 T2 formatting

```{r Clean T1 Bed files}
setwd("~/GWAS_22/new_gwas/T1T2")

#Load SNP info
bim<-fread("T1T2.bim")

#Filter SNPs in autosomal chromosomes")
non_auto <- bim %>% filter(V1 == "0" | V1 == "24")
#keep only SNPS containing rs IDs #
rs_bim <- bim %>% filter(!grepl("rs", V2))
#Remove duplicates 
duplicated_snps <- bim[duplicated(paste(bim$V1,bim$V4)),]

#remove snps where A > T, C > G
#PLINK can't tell what strand to merge/read these SNPs from
atcg_snps <-  bim[((bim$V5 == "A") & (bim$V6 == "T")) |
                         ((bim$V5 == "T") & (bim$V6 == "A")) |
                         ((bim$V5 == "C") & (bim$V6 == "G")) |
                         ((bim$V5 == "G") & (bim$V6 == "C"))]
#Exclusion list
exclude <- rbind(non_auto, duplicated_snps, atcg_snps, rs_bim)
exclude <- exclude$V2
write.table(file="exclude_snps_rm.txt", exclude, sep = "\t", quote=F, row.names = F, col.names = F)
system("plink --bfile T1T2 --exclude exclude_snps_rm.txt --make-bed --out T1T2_rm")
#error invalid centimorgan position on line 58721 of T1T2.bim.
#use Plink1.9 to convert for now, error reported
```


```{bash Make T1 Bed liftover files}
cd ~/GWAS_22/new_gwas/T1T2
awk '{print "chr" $1, $4 -1, $4, $2 }' T1T2_rm.bim > T1T2_rm_lift

#https://genome.ucsc.edu/cgi-bin/hgLiftOver
#Successfully converted 693434 records: View Conversions
#Conversion failed on 18369 records.
#Download as T1T2_38.bed
```


```{bash Update T1 with Liftover files}
cd ~/GWAS_22/new_gwas/T1T2
# extract mapped variants snp IDs
awk '{print $4}' T1T2_38.bed > T1T2_38_rm.snps
# extract updated positions
awk '{print $4, $3}' T1T2_38.bed > T1T2_38_rm.pos

#update plink file with SNPs that we could extract, and update the snp coordinates using update map function
plink2 --bfile T1T2_rm --make-pgen --out T1T2
plink2 --pfile T1T2 --extract T1T2_38_rm.snps --update-map T1T2_38_rm.pos --sort-vars --make-pgen --out T1T2_remap
plink2 --pfile T1T2_remap --make-bed --out T1T2_remap #bfile version
#Error: Invalid centimorgan position on line 57301 of T1T2_rm.bim.

```

```{r Update cm position}
#Some NaN values in .bim file, set cm value to dummy value of 0
#There are ways to calculate it however this doesn't seem essential for now
#Will try just changing NaN values to 0 first, if not will replace the whole col with 0's
setwd("~/GWAS_22/new_gwas/T1T2")
bim <- fread("T1T2_rm.bim")
bim$V3[grepl("NaN",bim$V3)] <-"0"
write.table(file="T1T2_rm.bim", bim, sep = "\t", quote=F, row.names = F, col.names = F)
#Rerun previous chunk with new bim file - worked :)
```

## P1 Tyger Formatting

```{r Clean P1 Bed files}
setwd("~/GWAS_22/new_gwas/P1Tyger")

#Load SNP info
bim<-fread("P1Tyger.bim")

#Filter SNPs in autosomal chromosomes")
non_auto <- bim %>% filter(V1 == "0" | V1 == "24")
#keep only SNPS containing rs IDs #
rs_bim <- bim %>% filter(!grepl("rs", V2))
#Remove duplicates 
duplicated_snps <- bim[duplicated(paste(bim$V1,bim$V4)),]

#remove snps where A > T, C > G
#PLINK can't tell what strand to merge/read these SNPs from
atcg_snps <-  bim[((bim$V5 == "A") & (bim$V6 == "T")) |
                         ((bim$V5 == "T") & (bim$V6 == "A")) |
                         ((bim$V5 == "C") & (bim$V6 == "G")) |
                         ((bim$V5 == "G") & (bim$V6 == "C"))]
#Exclusion list
exclude <- rbind(non_auto, duplicated_snps, atcg_snps, rs_bim)
exclude <- exclude$V2
write.table(file="exclude_snps_rm.txt", exclude, sep = "\t", quote=F, row.names = F, col.names = F)
system("plink --bfile P1Tyger --exclude exclude_snps_rm.txt --make-bed --out P1Tyger_rm")

```


```{bash Make P1 Bed liftover files}
cd ~/GWAS_22/new_gwas/P1Tyger
awk '{print "chr" $1, $4 -1, $4, $2 }' P1Tyger_rm.bim > P1Tyger_rm_lift

#https://genome.ucsc.edu/cgi-bin/hgLiftOver
#Successfully converted 693434 records: View Conversions
#Conversion failed on 18369 records.
#Download as P1Tyger_38.bed
```


```{bash Update P1 with Liftover files}
cd ~/GWAS_22/new_gwas/P1Tyger
# extract mapped variants snp IDs
awk '{print $4}' P1Tyger_38.bed > P1Tyger_38_rm.snps
# extract updated positions
awk '{print $4, $3}' P1Tyger_38.bed > P1Tyger_38_rm.pos

#update plink file with SNPs that we could extract, and update the snp coordinates using update map function
plink2 --bfile P1Tyger_rm --make-pgen --out P1Tyger
plink2 --pfile P1Tyger --extract P1Tyger_38_rm.snps --update-map P1Tyger_38_rm.pos --sort-vars --make-pgen --out P1Tyger_remap
plink2 --pfile P1Tyger_remap --make-bed --out P1Tyger_remap #bfile version
#no cm issues
```

```{r VAST file cleaning}
#Duplicates etc. still need to be removed can do from remap file to check for now

setwd("~/GWAS_22/new_gwas/Updated_SNPs")

#Load SNP info
bim <-fread("VAST_remap.bim")

#Filter SNPs in autosomal chromosomes")
non_auto <- bim %>% filter(V1 == "0" | V1 == "24")
#keep only SNPS containing rs IDs #
rs_bim <- bim %>% filter(!grepl("rs", V2))
#Remove duplicates 
duplicated_snps <- bim[duplicated(paste(bim$V1,bim$V4)),]

#remove snps where A > T, C > G
#PLINK can't tell what strand to merge/read these SNPs from
atcg_snps <-  bim[((bim$V5 == "A") & (bim$V6 == "T")) |
                         ((bim$V5 == "T") & (bim$V6 == "A")) |
                         ((bim$V5 == "C") & (bim$V6 == "G")) |
                         ((bim$V5 == "G") & (bim$V6 == "C"))]
#Exclusion list
exclude <- rbind(non_auto, duplicated_snps, atcg_snps, rs_bim)
exclude <- exclude$V2
write.table(file="exclude_snps_rm.txt", exclude, sep = "\t", quote=F, row.names = F, col.names = F)
system("plink --bfile VAST_remap --exclude exclude_snps_rm.txt --make-bed --out VAST_remap")
#only 612,676 variants left
```

## Alignment to reference genome
```{bash Alignment to ref genome}
cd ~/GWAS_22/new_gwas/Updated_SNPs
#Before we can merge/impute studies, we need to check for strand inconsistencies
#The best way to do this is to compare to a reference genome
#Plink offers a flip-scan option
#But will instead use the perl tool from Will Rayner
#Estimates possible flips by looking at overal LD, if suddenly a SNP is not in LD when it should be, very likely that that SNP has been incorrectly aligned

#generate TOPMED38 vcf.gz file to use

#VAST
#plink2 --bfile VAST_remap --freq --out VAST_remap
#perl pre_impute.pl -b VAST_remap.bim -f VAST_remap.afreq -r PASS.VariantsTOPMED38.tab.gz -h
#bash Run-plink.sh

#mv *VAST VAST/

#T1T2
plink2 --bfile T1T2_remap --freq --out T1T2_remap
perl pre_impute.pl -b T1T2_remap.bim -f T1T2_remap.afreq -r PASS.VariantsTOPMED38.tab.gz -h
bash Run-plink.sh
mv *T1T2* T1T2/

#P1Tyger
plink2 --bfile P1Tyger_remap --freq --out P1Tyger_remap
perl pre_impute.pl -b P1Tyger_remap.bim -f P1Tyger_remap.afreq -r PASS.VariantsTOPMED38.tab.gz -h
bash Run-plink.sh
mv *P1Tyger* P1Tyger/
```

```{bash Concat Chrs}
bcftools concat T1T2_remap-updated-chr{1..22}.vcf -Oz -o T1T2_concat.vcf.gz
bcftools concat P1Tyger_remap-updated-chr{1..22}.vcf -Oz -o P1Tyger_concat.vcf.gz
bcftools concat VAST_remap-updated-chr{1..22}.vcf -Oz -o VAST_concat.vcf.gz
```
```{bash Merge studies}
bcftools index VAST_concat.vcf.gz 
echo DONE index VAST 
bcftools index P1Tyger_concat.vcf.gz 
echo DONE index P1 
bcftools index T1T2_concat.vcf.gz 
echo DONE index T1 

#Tested merging of all 3 studies and not enough common snps between, even with updated GSA annotation
#Merged and QC'd T1 and P1 studies

bcftools merge T1T2_concat.vcf.gz P1Tyger_concat.vcf.gz -Oz -o P1T1.vcf.gz
plink2 --vcf P1T1.vcf.gz --max-alleles 2 --keep-allele-order --make-bed --out P1T1

plink2 --bfile P1T1 --geno 0.1 --mind 0.1 --make-bed --out P1T1_qc
plink2 --bfile P1T1_qc --geno 0.05 --make-bed --out P1T1_qc2

#Keep VAST separate

#Investigated difference in subsequent merge
#Of these, 437,406 are new, while 153,095 are present in the base dataset.
#Only 150,000 common SNPs between studies
#Therefore impute separately then merge

plink2 --vcf VAST_concat.vcf.gz --max-alleles 2 --keep-allele-order --make-bed --out VAST
plink2 --bfile VAST --geno 0.1 --mind 0.1 --make-bed --out VAST_qc
plink2 --bfile VAST_qc --geno 0.05 --make-bed --out VAST_qc2

#Update GSA annotation in r
r VP_bim$V2 <- str_replace(VP_bim$V2, "GSA-", "")
r write.table(file="VAST_qc2.bim", VP_bim, sep = "\t", quote=F, row.names = F, col.names = F)

#Recheck merge
plink --bfile VAST_qc2 --bmerge P1T1_qc2 --out merge

#See log files for more info

```
```{bash Format for TopMed}
bash topmed_prep.sh
```

